lib/benchmark/benchmark.go

```diff
@@ -149,157 +157,133 @@ func (c *Config) Benchmark(ctx context.Context, tc *client.TeleportClient) (Resu
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-	requestC := make(chan benchMeasure)
-	responseC := make(chan benchMeasure, c.Threads)
+	requestsC := make(chan benchMeasure)
+	resultC := make(chan benchMeasure)
 
-	for i := 0; i < c.Threads; i++ {
-		thread := &benchmarkThread{
-			id:          i,
-			ctx:         ctx,
-			client:      tc,
-			command:     c.Command,
-			interactive: c.Interactive,
-			receiveC:    requestC,
-			sendC:       responseC,
+	go func() {
+		interval := time.Duration(1 / float64(c.Rate) * float64(time.Second))
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
+		start := time.Now()
+		for {
+			select {
+			case <-ticker.C:
+				// ticker makes its first tick after the given duration, not immediately
+				// this sets the send measure ResponseStart time accurately
+				delay = delay + interval
+				t := start.Add(delay)
+				measure := benchMeasure{
+					ResponseStart: t,
+					command:       c.Command,
+					client:        tc,
+					ctx:           ctx,
+					interactive:   c.Interactive,
+				}
+				requestsC <- measure
+			case <-ctx.Done():
+				close(requestsC)
+				return
+			}
 		}
-		go thread.run()
-	}
+	}()
 
-	go produceMeasures(ctx, c.Rate, requestC)
+	go run(ctx, requestsC, resultC)
 
 	var result Result
-	// from one millisecond to 60000 milliseconds (minute) with 3 digits precision, refer to constants
 	result.Histogram = hdrhistogram.New(minValue, maxValue, significantFigures)
-	results := make([]benchMeasure, 0, c.MinimumMeasurements)
 	statusTicker := time.NewTicker(1 * time.Second)
 	timeElapsed := false
 	start := time.Now()
-
 	for {
 		if c.MinimumWindow <= time.Since(start) {
 			timeElapsed = true
 		}
 		select {
-		case measure := <-responseC:
-			result.Histogram.RecordValue(int64(measure.End.Sub(measure.Start) / time.Millisecond))
-			results = append(results, measure)
-			if timeElapsed && len(results) >= c.MinimumMeasurements {
-				cancelWorkers()
+		case measure := <-resultC:
+			result.Histogram.RecordValue(int64(measure.End.Sub(measure.ResponseStart) / time.Millisecond))
+			result.RequestsOriginated++
+			if timeElapsed && result.RequestsOriginated >= c.MinimumMeasurements {
+				cancel()
 			}
 			if measure.Error != nil {
 				result.RequestsFailed++
 				result.LastError = measure.Error
 			}
-			result.RequestsOriginated++
 		case <-ctx.Done():
 			result.Duration = time.Since(start)
 			return result, nil
 		case <-statusTicker.C:
-			logrus.Infof("working... observations: %d", len(results))
+			logrus.Infof("working... current observation count: %d", result.RequestsOriginated)
 		}
+
 	}
 }
 
-func produceMeasures(ctx context.Context, rate int, c chan<- benchMeasure) {
-	interval := time.Duration(1 / float64(rate) * float64(time.Second))
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
+type benchMeasure struct {
+	ResponseStart time.Time
+	End           time.Time
+	Error         error
+	ctx           context.Context
```

Comment:
this seems to be unused
and context should be always passed as a separate parameter: https://github.com/golang/go/wiki/CodeReviewComments#contexts